pkg/rke2/spw.go

```diff
@@ -0,0 +1,104 @@
+package rke2
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	containerdk3s "github.com/k3s-io/k3s/pkg/agent/containerd"
+	"github.com/k3s-io/k3s/pkg/cli/cmds"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
+)
+
+// total of 4 minutes
+var podCheckBackoff = wait.Backoff{
+	Steps:    12,
+	Duration: 15 * time.Second,
+	Factor:   1.0,
+	Jitter:   0.1,
+}
+
+// total of 31 seconds
+var criBackoff = wait.Backoff{
+	Steps:    5,
+	Duration: 1 * time.Second,
+	Factor:   2,
+	Jitter:   0.1,
+}
+
+// checkStaticManifests validates that the pods started with rke2 match the static manifests
+// provided in /var/lib/rancher/rke2/agent/pod-manifests. When restarting rke2, it takes time
+// for any changes to static manifests to be pulled by kubelet. Additionally this prevents errors
+// where something is wrong with the static manifests and RKE2 starts anyways.
+func checkStaticManifests(dataDir string) cmds.StartupHook {
+	return func(ctx context.Context, wg *sync.WaitGroup, args cmds.StartupHookArgs) error {
+		go func() {
+			defer wg.Done()
+			<-args.APIServerReady
```

Comment:
Do we need to wait for this? We're only talking to containerd, and the apiserver manifest is one of the things we're examining, so waiting for it to start first is counterintuitive to me.

Reply:
Removed, this was leftover from all the other startup hooks that I used as a base for this one.